### ISSUE TITLE
Deserialization panics in structs with two custom enums

### ISSUE THREAD

**Issue by SuperFluffy** [3Hren/msgpack-rust#92]
When trying to deserialize a struct containing two enums, the Deserializer panics. If I put the enums into a struct by themselves, everything works fine. But the moment they are combined into one, all hell breaks loose. See the example below. Any idea what's going on?

```rust
#![feature(proc_macro)]

extern crate rmp_serde;
extern crate serde;

#[macro_use]
extern crate serde_derive;

use serde::{Deserialize,Serialize};
use rmp_serde::Serializer;

#[derive(Debug, Deserialize, Serialize)]
pub enum Foo {
    A, 
    B,
}

#[derive(Debug, Deserialize, Serialize)]
pub enum Bar {
    C,
    D,
}

use Foo::*;
use Bar::*;

#[derive(Debug,Deserialize,Serialize)]
pub struct Combined {
    pub foo: Foo,
    pub bar: Bar,
}

#[derive(Debug,Deserialize,Serialize)]
pub struct JustFoo {
    pub foo: Foo,
}

#[derive(Debug,Deserialize,Serialize)]
pub struct JustBar {
    pub bar: Bar,
}

#[derive(Debug,Deserialize,Serialize)]
pub struct JustCombined {
    pub foo: JustFoo,
    pub bar: JustBar,
}

fn main() {
    let c = Combined {
        foo: A,
        bar: C,
    };

    let jf = JustFoo {
        foo: A,
    };

    let jb = JustBar {
        bar: C,
    };

    let jc = JustCombined {
        foo: JustFoo{ foo: A },
        bar: JustBar{ bar: C },
    };

    // ERROR:
    // thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: TypeMismatch(Null)', ../src/libcore/result.rs:837
    // let mut buffer = Vec::new();
    // Serialize::serialize(&c, &mut Serializer::new(&mut buffer)).ok().unwrap();
    // let mut deserializer = rmp_serde::Deserializer::new(&buffer[..]);
    // let c_de: Combined = Deserialize::deserialize(&mut deserializer).unwrap();

    let mut buffer = Vec::new();
    Serialize::serialize(&jf, &mut Serializer::new(&mut buffer)).ok().unwrap();
    let mut deserializer = rmp_serde::Deserializer::new(&buffer[..]);
    let jf_de: JustFoo = Deserialize::deserialize(&mut deserializer).unwrap();

    let mut buffer = Vec::new();
    Serialize::serialize(&jb, &mut Serializer::new(&mut buffer)).ok().unwrap();
    let mut deserializer = rmp_serde::Deserializer::new(&buffer[..]);
    let jb_de: JustBar = Deserialize::deserialize(&mut deserializer).unwrap();

    // ERROR:
    // thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: TypeMismatch(Null)', ../src/libcore/result.rs:837
    // let mut buffer = Vec::new();
    // Serialize::serialize(&jc, &mut Serializer::new(&mut buffer)).ok().unwrap();
    // let mut deserializer = rmp_serde::Deserializer::new(&buffer[..]);
    // let jc_de: JustCombined = Deserialize::deserialize(&mut deserializer).unwrap();
}
```

**Comment by SuperFluffy** [comment]
The equivalent `bincode` example works fine:

```rust
#![feature(proc_macro)]

extern crate bincode;
extern crate ndarray;
extern crate serde;

#[macro_use]
extern crate serde_derive;

use bincode::SizeLimit::Infinite;
use serde::{Deserialize,Serialize};
use std::fs::File;
use std::io::BufWriter;
use std::io::prelude::*;

#[derive(Debug, Deserialize, Serialize)]
pub enum Foo {
    A, 
    B,
}

#[derive(Debug, Deserialize, Serialize)]
pub enum Bar {
    C,
    D,
}

use Foo::*;
use Bar::*;

#[derive(Debug,Deserialize,Serialize)]
pub struct Combined {
    pub foo: Foo,
    pub bar: Bar,
}

#[derive(Debug,Deserialize,Serialize)]
pub struct JustFoo {
    pub foo: Foo,
}

#[derive(Debug,Deserialize,Serialize)]
pub struct JustBar {
    pub bar: Bar,
}

#[derive(Debug,Deserialize,Serialize)]
pub struct JustCombined {
    pub foo: JustFoo,
    pub bar: JustBar,
}

fn main() {
    let c = Combined {
        foo: A,
        bar: C,
    };

    let jf = JustFoo {
        foo: A,
    };

    let jb = JustBar {
        bar: C,
    };

    let jc = JustCombined {
        foo: JustFoo{ foo: A },
        bar: JustBar{ bar: C },
    };

    let mut buf = bincode::serde::serialize(&c, Infinite).unwrap();
    let c_de: Combined = bincode::serde::deserialize(&buf).unwrap();
    println!("{:?} -> {:?}", c, c_de);

    let mut buf = bincode::serde::serialize(&jf, Infinite).unwrap();
    let jf_de: JustFoo = bincode::serde::deserialize(&buf).unwrap();
    println!("{:?} -> {:?}", jf, jf_de);

    let mut buf = bincode::serde::serialize(&jb, Infinite).unwrap();
    let jb_de: JustBar = bincode::serde::deserialize(&buf).unwrap();
    println!("{:?} -> {:?}", jb, jb_de);

    let mut buf = bincode::serde::serialize(&jc, Infinite).unwrap();
    let jc_de: JustCombined = bincode::serde::deserialize(&buf).unwrap();
    println!("{:?} -> {:?}", jc, jc_de);
}
```

**Comment by knsd** [comment]
Fixed in eea556a28452bfd58000b174d05964495e89f4a1.

**Comment by SuperFluffy** [comment]
This is indeed fixed. Thank you!
